Peripheral definitions in cortex-m are not `Send`, while those from `svd2rust` generated crates are

Comparing something like:
 - [`stm32f103xx::TIM6`](https://docs.rs/stm32f103xx/0.8.0/stm32f103xx/struct.TIM6.html)
 - [`cortex_m::ITM`](https://docs.rs/cortex-m/0.4.1/cortex_m/peripheral/struct.ITM.html)

Notably:
 - `Send` is omitted from `cortex_m::ITM`
 - mutable access (in the form of `ptr() -> *mut RegisterBlock` and `DerefMut` are removed from `stm32f103xx::TIM6`

I ran into this while trying to pass around `ITM` in `cortex_m_rtfm`, which wants `Send` on `resources`:

```
error[E0277]: the trait bound `*const (): core::marker::Send` is not satisfied in `cortex_m::peripheral::ITM`
  --> src/main.rs:16:1
   |
16 | / app! {
17 | |     device: stm32f429,
18 | |
19 | |     resources: {
...  |
29 | |     },
30 | | }
   | |_^ `*const ()` cannot be sent between threads safely
   |
   = help: within `cortex_m::peripheral::ITM`, the trait `core::marker::Send` is not implemented for `*const ()`
   = note: required because it appears within the type `core::marker::PhantomData<*const ()>`
   = note: required because it appears within the type `cortex_m::peripheral::ITM`
```

I've used a quick wrapper to workaround:
```
...
use core::ops::{Deref,DerefMut};
pub struct ITM(cortex_m::peripheral::ITM);

impl Deref for ITM {
    type Target = cortex_m::peripheral::ITM;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for ITM {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}
unsafe impl Send for ITM {}

...
app! {
...
    resources: {
        static R_ITM: ITM;
        ...
    },

    tasks: {
        TIM7: {
            path: periodic,
            resources: [R_ITM, ...],
        },
    },
}

...
fn init(mut p: init::Peripherals) -> init::LateResources {
    ...

    init::LateResources {
        R_ITM: ITM(p.core.ITM),
        R_TIM7: tim,
    }
}

...
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,12 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased]
 
+## [v0.4.2] - 2018-01-17
+
+### Fixed
+
+- Added a missing `Send` implementation to all the peripherals.
+
 ## [v0.4.1] - 2018-01-16
 
 ### Changed
@@ -385,7 +391,8 @@ fn main() {
 - Functions to get the vector table
 - Wrappers over miscellaneous instructions like `bkpt`
 
-[Unreleased]: https://github.com/japaric/cortex-m/compare/v0.4.1...HEAD
+[Unreleased]: https://github.com/japaric/cortex-m/compare/v0.4.2...HEAD
+[v0.4.2]: https://github.com/japaric/cortex-m/compare/v0.4.1...v0.4.2
 [v0.4.1]: https://github.com/japaric/cortex-m/compare/v0.4.0...v0.4.1
 [v0.4.0]: https://github.com/japaric/cortex-m/compare/v0.3.1...v0.4.0
 [v0.3.1]: https://github.com/japaric/cortex-m/compare/v0.3.0...v0.3.1

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -7,7 +7,7 @@ keywords = ["arm", "cortex-m", "register", "peripheral"]
 license = "MIT OR Apache-2.0"
 name = "cortex-m"
 repository = "https://github.com/japaric/cortex-m"
-version = "0.4.1"
+version = "0.4.2"
 
 [dependencies]
 aligned = "0.1.1"

--- a/src/peripheral/mod.rs
+++ b/src/peripheral/mod.rs
@@ -221,6 +221,9 @@ pub struct CBP {
 }
 
 #[cfg(any(armv7m, target_arch = "x86_64"))]
+unsafe impl Send for CBP {}
+
+#[cfg(any(armv7m, target_arch = "x86_64"))]
 impl CBP {
     pub(crate) unsafe fn new() -> Self {
         CBP {
@@ -235,9 +238,6 @@ impl CBP {
 }
 
 #[cfg(any(armv7m, target_arch = "x86_64"))]
-unsafe impl Send for CBP {}
-
-#[cfg(any(armv7m, target_arch = "x86_64"))]
 impl ops::Deref for CBP {
     type Target = self::cbp::RegisterBlock;
 
@@ -250,6 +250,8 @@ impl ops::Deref for CBP {
 pub struct CPUID {
     _marker: PhantomData<*const ()>,
 }
+
+unsafe impl Send for CPUID {}
 
 impl CPUID {
     /// Returns a pointer to the register block
@@ -271,6 +273,8 @@ pub struct DCB {
     _marker: PhantomData<*const ()>,
 }
 
+unsafe impl Send for DCB {}
+
 impl DCB {
     /// Returns a pointer to the register block
     pub fn ptr() -> *const dcb::RegisterBlock {
@@ -290,6 +294,8 @@ impl ops::Deref for DCB {
 pub struct DWT {
     _marker: PhantomData<*const ()>,
 }
+
+unsafe impl Send for DWT {}
 
 impl DWT {
     /// Returns a pointer to the register block
@@ -313,6 +319,9 @@ impl ops::Deref for DWT {
 pub struct FPB {
     _marker: PhantomData<*const ()>,
 }
+
+#[cfg(any(armv7m, target_arch = "x86_64"))]
+unsafe impl Send for FPB {}
 
 #[cfg(any(armv7m, target_arch = "x86_64"))]
 impl FPB {
@@ -340,6 +349,9 @@ pub struct FPU {
 }
 
 #[cfg(any(has_fpu, target_arch = "x86_64"))]
+unsafe impl Send for FPU {}
+
+#[cfg(any(has_fpu, target_arch = "x86_64"))]
 impl FPU {
     /// Returns a pointer to the register block
     pub fn ptr() -> *const fpu::RegisterBlock {
@@ -363,6 +375,9 @@ impl ops::Deref for FPU {
 pub struct ITM {
     _marker: PhantomData<*const ()>,
 }
+
+#[cfg(any(armv7m, target_arch = "x86_64"))]
+unsafe impl Send for ITM {}
 
 #[cfg(any(armv7m, target_arch = "x86_64"))]
 impl ITM {
@@ -393,6 +408,8 @@ pub struct MPU {
     _marker: PhantomData<*const ()>,
 }
 
+unsafe impl Send for MPU {}
+
 impl MPU {
     /// Returns a pointer to the register block
     pub fn ptr() -> *const mpu::RegisterBlock {
@@ -412,6 +429,8 @@ impl ops::Deref for MPU {
 pub struct NVIC {
     _marker: PhantomData<*const ()>,
 }
+
+unsafe impl Send for NVIC {}
 
 impl NVIC {
     /// Returns a pointer to the register block
@@ -433,6 +452,8 @@ pub struct SCB {
     _marker: PhantomData<*const ()>,
 }
 
+unsafe impl Send for SCB {}
+
 impl SCB {
     /// Returns a pointer to the register block
     pub fn ptr() -> *const scb::RegisterBlock {
@@ -452,6 +473,8 @@ impl ops::Deref for SCB {
 pub struct SYST {
     _marker: PhantomData<*const ()>,
 }
+
+unsafe impl Send for SYST {}
 
 impl SYST {
     /// Returns a pointer to the register block
@@ -475,6 +498,9 @@ impl ops::Deref for SYST {
 pub struct TPIU {
     _marker: PhantomData<*const ()>,
 }
+
+#[cfg(any(armv7m, target_arch = "x86_64"))]
+unsafe impl Send for TPIU {}
 
 #[cfg(any(armv7m, target_arch = "x86_64"))]
 impl TPIU {

